Add `scope` for aliased delegate function parameters too

mihails-strasuns-sociomantic opened the pull request:
Implemented by adding dependency on `dsymbol` library providing
basic semantic analysis of D sources and doing lookup of symbols
when visitin function parameters.

mihails-strasuns-sociomantic committed:
Stop converting alias declaration

Declaring delegate alias as `scope` doesn't propagate to
function argument when such alias is used as parameter and
thus is nothing but noise.

mihails-strasuns-sociomantic committed:
Integrate dsymbol to track all delegate aliases

Fixes #1

Improves d1to2fix capability of marking all delegate arguments
as scope by keeping cache of all aliases that resolve to
delegates in converted sources.

* submodules/dsymbol ()...59d1c26(59d1c26) (1 commits)
  > Merge pull request #11 from WebFreak001/master

* submodules/emsi_containers ()...3c6a439(3c6a439) (1 commits)
  > #49

mihails-strasuns-sociomantic committed:
Change test runner to convert all snippets at once

With addition of semantic phase d1to2fix startup time has become
too slow to conveniently convert files one by one as it results
in redundant building of module cache each time.

mihails-strasuns-sociomantic commented:
This is in minimal working implementation now, will definitely need some refactoring before merging. @mathias-lang-sociomantic I am a bit afraid to touch your delegate visiting logic, maybe you can suggest best way to merge it with added code? :)

@leandro-lucarella-sociomantic FYI this will require adjustments in makd too:

1) it must provide list of all files to convert at once as startup of d1to2fix is rather slow now (few seconds with only `/usr/include/d2/dmd-transitional` being cached)
2) it must pass all `-I` flags to `d1to2fix` same as it does for `DFLAGS`

mihails-strasuns-sociomantic commented:
Worth noting that right now symbol search code may looks a bit more complicated than necessary because it was adapted from DCD code where it served somewhat different purpose ("go to declaration" hint implementation). Contributing more efficient API upstream is desired but out of this PR scope for now.

mathias-lang-sociomantic commented on src/d1to2fix/symbolsearch.d:
```diff
@@ -0,0 +1,80 @@
+	const ModuleCache* moduleCache_)
+{
+    // until https://github.com/Hackerpilot/dsymbol/pull/12 is merged
+    auto moduleCache = cast(ModuleCache*) moduleCache_;
```
It was

mathias-lang-sociomantic commented on src/d1to2fix/main.d:
```diff
@@ -102,6 +110,17 @@ int main ( string[] args )
+    // Build symbol cache for parsed modules
+
+    import d1to2fix.symbolsearch;
+    config.importPaths ~= "/usr/include/d2/dmd-transitional/";
```
This is quite specific

leandro-lucarella-sociomantic replied:
Pending.

mihails-strasuns-sociomantic commented:
Have been testing it with ocean (via manual `d1to2fix @modules.txt`) - things keep working as they are when module cache is empty (no import paths) but adding `-I./src/` breaks stuff because of interesting false positives. This issue has proven much more trouble than I expected but I am still going to pursue it to the end as relevant experience will later help a lot with our D2 tooling implementation.

mihails-strasuns-sociomantic commented:
Ok, after moving `scope` token injection into beginning of the type ocean keeps passing tests, now playing with node.

mihails-strasuns-sociomantic commented:
After reimplementing symbol cache logic it now converts node properly and results in huge improvement of GC collection count and total GC time. Now making code pretty ..

mihails-strasuns-sociomantic committed:
Documentation & cleanup

mihails-strasuns-sociomantic commented:
Updated:
- https://github.com/sociomantic-tsunami/d1to2fix/pull/4 split into own PR
- Added documentation/comments
- Minor refactoring of visitor internals

Should be ready for review

mihails-strasuns-sociomantic commented:
Ping @leandro-lucarella-sociomantic @mathias-lang-sociomantic

mathias-lang-sociomantic commented on src/d1to2fix/symbolsearch.d:
I guess that is because of:

``` D
class A
{
 A field;
}
```

?

mihails-strasuns-sociomantic commented on src/d1to2fix/symbolsearch.d:
Possibly - or something similar with `alias this`. I haven't investigated it as module scope proved enough for dhtnode goals.

leandro-lucarella-sociomantic commented on src/d1to2fix/main.d:
```diff
@@ -102,6 +110,20 @@ int main ( string[] args )
+    initializeModuleCache(config.importPaths);
+
+    import std.stdio;
+    writefln("Module cache initialized for %s", config.importPaths);
```
Not using the `logger` on purpose here?

mihails-strasuns-sociomantic replied:
Left-over trace from checking parse timings, will remove

leandro-lucarella-sociomantic commented on src/d1to2fix/symbolsearch.d:
I really dislike this style. IMHO the function scope is small enough to be clear to declare all at the top of the function, instead of polluting all the logic with `import` statements, and it will also dirty the diffs once you need to use some symbol in these imports in some code above this declaration, as you'll have to move imports around too.

mihails-strasuns-sociomantic commented on src/d1to2fix/symbolsearch.d:
Practice has shown that imports from function top level are never getting removed when not used anymore, contrary to using them like here.

mihails-strasuns-sociomantic commented on src/d1to2fix/symbolsearch.d:
I also like Carmack view that such code blocks can be considered as nested force-inlined functions and reasoned in a similar fashion (http://number-none.com/blow/john_carmack_on_inlined_code.html)

leandro-lucarella-sociomantic commented on src/d1to2fix/symbolsearch.d:
I think you can always forget to remove those imports, specially if they are used in code below the declaration already. In any case, I think looking for unused imports is a job for a lint tool, it doesn't make to do so manually, so I think adding of the noise to the logic is justified. But I don't want to bomb this PR with the off-topic, is a discussion for another time and place... :)

leandro-lucarella-sociomantic commented on src/d1to2fix/symbolsearch.d:
```diff
@@ -0,0 +1,70 @@
+    NB: must be called in main thread before any file conversion workers
+    start to avoid any race conditions
+ */
+public void initializeModuleCache(string[] paths)
```
Shoudn't this be put in a module constructor then?

leandro-lucarella-sociomantic replied:
Also pending.

leandro-lucarella-sociomantic commented:
Besides the note about imports and not being too familiar with d1to2fix internals, LGTM.

mihails-strasuns-sociomantic commented on src/d1to2fix/main.d:
```diff
@@ -102,6 +110,17 @@ int main ( string[] args )
+    // Build symbol cache for parsed modules
+
+    import d1to2fix.symbolsearch;
+    config.importPaths ~= "/usr/include/d2/dmd-transitional/";
```
I can't imagine any other place where to put it unless we want `d1to2fix` to parse `dmd.conf` - makd can't provide it either.

leandro-lucarella-sociomantic replied:
Fair enough.

leandro-lucarella-sociomantic replied:
But maybe add a comment then?

mihails-strasuns-sociomantic commented:
Updated:
- removed left-over trace print

mihails-strasuns-sociomantic commented:
Ping. @mathias-lang-sociomantic could you try this out in your app one last time to sanity check backwards compatibility?

mathias-lang-sociomantic commented:
Sure. Did you get an estimate on how much more coverage it adds / how many `delegate` are still missed ?

mihails-strasuns commented:
For dhtnode it seems to conver 100% of previously allocated closures - at least during my performance tests D2 version have never caused GC collection cycle under same load (identical to D1 build). Haven't checked generic impact.

NB: if you just build new d1to2fix, it won't convert any new delegates, you need to also pass same `-I` flag set as used by compiler for that.

mathias-lang-sociomantic commented:
Looks borken again (14.04):

```
~/tsunami/d1to2fix (git)-[fc1bb06...] % make clean; make all test
   rm build
   dmd build/devel/bin/libstdextra.a
   rdmd build/devel/bin/d1to2fix
   gen build/devel/tmp/allunittests.d
   rdmd build/devel/tmp/allunittests
   run build/devel/tmp/allunittests
Running unittest for unDecorateComment...
Unittest for unDecorateComment passed.
   rdmd build/devel/tmp/test-snippets
   run build/devel/tmp/test-snippets
object.Exception@/home/geod24/tsunami/d1to2fix/test/snippets/main.d(20): ./build/last/bin/d1to2fix -I./tests/ ./tests/this.d ./tests/issue21.d ./tests/aliasdg.d ./tests/issue22.d ./tests/scope_delegate.d ./tests/injection.d ./tests/const.d
----------------
/usr/include/dmd/phobos/std/exception.d:405 pure @safe void std.exception.bailOut!(Exception).bailOut(immutable(char)[], ulong, const(char[])) [0x4a685b]
/usr/include/dmd/phobos/std/exception.d:363 pure @safe bool std.exception.enforce!(Exception, bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) [0x4a67de]
/home/geod24/tsunami/d1to2fix/test/snippets/main.d:20 _Dmain [0x4a189e]
??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x4bb51a]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x4bb464]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x4bb4d6]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x4bb464]
??:? _d_run_main [0x4bb3d5]
??:? main [0x4b8a53]
??:? __libc_start_main [0xf9c50f44]
```

In particular:

```
~/tsunami/d1to2fix (git)-[fc1bb06...] % ./build/last/bin/d1to2fix -I./tests/ ./tests/this.d ./tests/issue21.d ./tests/aliasdg.d ./tests/issue22.d ./tests/scope_delegate.d ./tests/injection.d ./tests/const.d                                                                                          :(
object.Exception@/usr/include/dmd/phobos/std/stdio.d(816): rawRead must take a non-empty buffer
----------------
/usr/include/dmd/phobos/std/stdio.d:816 ubyte[] std.stdio.File.rawRead!(ubyte).rawRead(ubyte[]) [0x761e84]
/home/geod24/tsunami/d1to2fix/src/d1to2fix/main.d:137 int d1to2fix.main.main(immutable(char)[][]).__foreachbody2(ref immutable(char)[]) [0x75c295]
/usr/include/dmd/phobos/std/parallelism.d-mixin-3785:3829 _D3std11parallelism26__T15ParallelForeachTAAyaZ15ParallelForeach7opApplyMFMDFKAyaZiZ4doItMFZv [0x761913]
??:? void std.parallelism.run!(void delegate()).run(void delegate()) [0x8af467]
??:? void std.parallelism.__T4TaskS213std11parallelism3runTDFZvZ.Task.impl(void*) [0x8aefa7]
??:? void std.parallelism.AbstractTask.job() [0x8dc2ba]
??:? void std.parallelism.TaskPool.doJob(std.parallelism.AbstractTask*) [0x8adc6b]
??:? void std.parallelism.TaskPool.executeWorkLoop() [0x8addbd]
??:? void std.parallelism.TaskPool.startWorkLoop() [0x8add65]
??:? void core.thread.Thread.run() [0x8e0c35]
??:? thread_entryPoint [0x8e0a1f]
??:? [0x16b49183]
```

mathias-lang-sociomantic commented:
Oh, actually, it only crashes on the second run (which shouldn't happen)

mihails-strasuns-sociomantic commented:
Crash is not good but it is indeed expect to fail on second time because it converts test files in place now - I can't use stdout when converting all at once :(

mathias-lang-sociomantic commented:
Actually what happens is funnier.

First run -> Succeed

Second run:

> ./tests/injection.d(4:14)[error]: 'in', 'out', 'body', or block statement expected
> 1 parse errors encountered. Aborting upgrade of ./tests/injection.d

The the whole content of `tests/injection.d` is lost and we're left with an empty file. I'm not sure how I feel about this, since my workflow included `d1to2fix` on WIP files on multiple occasions.

Third and later runs:
`injection.d` is empty, so we're creating an unitialized array of size '0', which is passed to `rawRead` and it fails.

The third cause of failure should really be fixed IMO, if someone has an empty file hanging it its source it should not crash `d1to2fix`. And if we could avoid trashing the whole file when some error happened, that'd be great.

mathias-lang-sociomantic commented:
I tested some libraries individually, `p*******r`, and 3 applications belonging to my teams.
What are the odds that it _only_ crash on those 3 apps ?

```
gdb --args d1to2fix -I/home/geod24/projects/myapp/build/devel/include -I./src  -I./submodules/makd/src  -I./submodules/ocean/src  -I./submodules/swarm/src  -I./submodules/library1/src  -I./submodules/yourlib/src  -I./submodules/yourlibdependency/src `git ls-files | grep '\.d$' | tr '\n' ' '`
```

```
Program received signal SIGSEGV, Segmentation fault.
0x000000000085b51d in containers.ttree.TTree!(dsymbol.symbol.SymbolOwnership, std.experimental.allocator.mallocator.Mallocator, true, "a < b", false, 64uL).TTree.range!(containers.ttree.TTree!(dsymbol.symbol.SymbolOwnership, std.experimental.allocator.mallocator.Mallocator, true, "a < b", false, 64uL).TTree).range() (this=..., __HID256=0x7fffffffbfc0) at ./submodules/emsi_containers/src/containers/ttree.d:174
174         return Range!(This)(cast(const(Node)*) root, RangeType.all, T.init);
(gdb) p root
Cannot access memory at address 0x7ffff7eb5020
(gdb) bt
#0  0x000000000085b51d in containers.ttree.TTree!(dsymbol.symbol.SymbolOwnership, std.experimental.allocator.mallocator.Mallocator, true, "a < b", false, 64uL).TTree.range!(containers.ttree.TTree!(dsymbol.symbol.SymbolOwnership, std.experimental.allocator.mallocator.Mallocator, true, "a < b", false, 64uL).TTree).range() (this=..., __HID256=0x7fffffffbfc0) at ./submodules/emsi_containers/src/containers/ttree.d:174
#1  0x000000000080d415 in dsymbol.symbol.DSymbol.~this() (this=...) at ./submodules/dsymbol/src/dsymbol/symbol.d:190
#2  0x000000000080dca1 in dsymbol.symbol.DSymbol.__aggrDtor() (this=...) at ./submodules/dsymbol/src/dsymbol/symbol.d:188
#3  0x0000000000894415 in object.TypeInfo_Struct.destroy(void*) const ()
#4  0x00000000007feafb in dsymbol.cache_entry.CacheEntry.~this() (this=...) at ./submodules/dsymbol/src/dsymbol/cache_entry.d:28
#5  0x00000000007fecf9 in dsymbol.cache_entry.CacheEntry.__aggrDtor() (this=...) at ./submodules/dsymbol/src/dsymbol/cache_entry.d:25
#6  0x00000000007699b1 in object._destructRecurse!(dsymbol.cache_entry.CacheEntry)._destructRecurse(ref dsymbol.cache_entry.CacheEntry) (s=...) at /usr/include/dmd/druntime/import/object.d:2409
#7  0x0000000000769911 in object.destroy!(dsymbol.cache_entry.CacheEntry).destroy(ref dsymbol.cache_entry.CacheEntry) (obj=...) at /usr/include/dmd/druntime/import/object.d:2778
#8  0x00000000007f856d in std.experimental.allocator.dispose!(shared(std.experimental.allocator.mallocator.Mallocator), dsymbol.cache_entry.CacheEntry).dispose(ref shared(std.experimental.allocator.mallocator.Mallocator), dsymbol.cache_entry.CacheEntry*) (p=0xc55a80, alloc=...)
    at ./submodules/experimental_allocator/src/std/experimental/allocator/package.d:1057
#9  0x00000000008694d0 in dsymbol.modulecache.ModuleCache.~this() (this=...) at ./submodules/dsymbol/src/dsymbol/modulecache.d:81
#10 0x000000000086add9 in dsymbol.modulecache.ModuleCache.__aggrDtor() (this=...) at ./submodules/dsymbol/src/dsymbol/modulecache.d:79
#11 0x00000000008e39fa in rt.lifetime.finalize_struct(void*, ulong) ()
#12 0x00000000008d47f3 in rt_finalizeFromGC ()
#13 0x00000000008cf4e8 in gc.gc.Gcx.sweep() ()
#14 0x00000000008cfdd5 in gc.gc.Gcx.fullcollect(bool) ()
#15 0x00000000008d2475 in gc.gc.GC.runLocked!(gc.gc.GC.fullCollectNoStack().go(gc.gc.Gcx*), gc.gc.Gcx*).runLocked(ref gc.gc.Gcx*) ()
#16 0x00000000008cc681 in gc.gc.GC.fullCollectNoStack() ()
#17 0x00000000008e322e in gc_term ()
#18 0x00000000008d3b94 in rt_term ()
#19 0x0000000000897b55 in rt.dmain2._d_run_main(int, char**, extern(C) int(char[][]) function*).runAll() ()
#20 0x0000000000897ad5 in rt.dmain2._d_run_main(int, char**, extern(C) int(char[][]) function*).tryExec(scope void() delegate) ()
#21 0x0000000000897a46 in _d_run_main ()
#22 0x000000000076ba6c in main ()
#23 0x00007ffff6eeff45 in __libc_start_main (main=0x76ba4c <main>, argc=55, argv=0x7fffffffd4a8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd498) at libc-start.c:287
#24 0x000000000075bd99 in _start ()
```

From a quick look it looks like it crashes on finalization. On some apps the code seems to be completely converted, on some others some files just disappear.

mihails-strasuns-sociomantic committed:
Close input file even if error is thrown

mihails-strasuns-sociomantic committed:
Don't try converting empty files

mihails-strasuns-sociomantic commented:
Pushed two fixes for file content thrashing and 0 length file crash. Looking into real project crash now.

mihails-strasuns-sociomantic commented:
Added some fixes and also changed tests to output converted code into new files to make it re-entrant. @mathias-lang-sociomantic can you test again? (your projects are converted fine on my machine)

mihails-strasuns-sociomantic committed:
Possible fix to finalization crashes with dsymbol

Prevents collecting of module cache by GC

mathias-lang-sociomantic commented:
No more crash.
The lack of recursion in aggregate is severely limiting its usefulness, but we can address this later.

mihails-strasuns-sociomantic commented:
@nemanja-boric-sociomantic @leandro-lucarella-sociomantic it is in your hands now

leandro-lucarella-sociomantic commented on src/d1to2fix/symbolsearch.d:
```diff
@@ -1,27 +1,38 @@
+    Imports
+
+******************************************************************************/
+
```
Pfff, I would like to eventually get rid of this imports header, is utterly useless.

leandro-lucarella-sociomantic replied:
We decided to kill them, so feel free to do so :)

leandro-lucarella-sociomantic commented on src/d1to2fix/visitor.d:
```diff
@@ -169,10 +169,63 @@ public final class TokenMappingVisitor : ASTVisitor
+        Params:
+            type = parameter type
+            token = parameter name token
+    **/
```
Why the different style in the comments (no infinite `****`)?

mihails-strasuns-sociomantic replied:
I have been using Phobos style for comments in d1to2fix originally but forgot about it when adding new stuff. Which way would you prefer to unify it?

leandro-lucarella-sociomantic replied:
You code it for the most part, so I think it makes more sense for you to decide. As long as it is consistent... :)

leandro-lucarella-sociomantic commented on src/d1to2fix/symbolsearch.d:
```diff
@@ -38,7 +38,10 @@ public void initializeModuleCache(string[] paths)
 
-    auto module_cache = new ModuleCache(new ASTAllocator);
+    // static is used so that pointer persists and GC won't collect it
+    // together with underlying owned dsymbols
```
Is there an issue for this?

mihails-strasuns-sociomantic replied:
No, this is "work as intended" domain of issues. Most of EMSI containers implement owning semantics and manually allocate memory internally. It is not designed to be used with GC like in old code.

leandro-lucarella-sociomantic commented:
@mathias-lang-sociomantic: @mihails-strasuns-sociomantic is on vacations for more than a month, I only made very minor comments, so if this is needed, I can merge as is and he can address the comments when he is back. Otherwise I would wait for him.

mihails-strasuns-sociomantic committed:
Unify DDOC style across modules

mihails-strasuns-sociomantic committed:
Explain why stdlib path is hard-coded

mihails-strasuns-sociomantic commented:
Updated with unified comment style.

leandro-lucarella-sociomantic commented on src/d1to2fix/symbolsearch.d:
```diff
@@ -0,0 +1,121 @@
+
+    License: Boost Software License Version 1.0 (see LICENSE for details)
+
+******************************************************************************/
```
This header still have a `**********` style comment, but I guess I can live with that :)

mihails-strasuns-sociomantic replied:
I kept all headers like that, it is intentional/consistent :)

leandro-lucarella-sociomantic replied:
Consistent inconsistency FTW!
